Ldap auth engine

erlisb opened the pull request:
This PR contains objects for new auth method, type LDAPAuthEngineConfig : 

- Create the CRD type and a validating webhooks for LDAPAuthEngineConfig
- Update type specs and webhook
- Generate CRD manifests
- Add test for LDAPAuthEngineConfig
- Add LDAPAuthEngineConfig section in readme

raffaelespazzoli commented on api/v1alpha1/ldapauthengineconfig_webhook.go:
```diff
@@ -0,0 +1,78 @@
+		Complete()
+}
+
+//+kubebuilder:webhook:path=/mutate-redhatcop-redhat-io-v1alpha1-ldapauthengineconfig,mutating=true,failurePolicy=fail,sideEffects=None,groups=redhatcop.redhat.io,resources=ldapauthengineconfigs,verbs=create;update,versions=v1alpha1,name=mldapauthengineconfig.kb.io,admissionReviewVersions=v1
```
please remove the update event here

erlisb replied:
removed.

raffaelespazzoli commented on api/v1alpha1/ldapauthengineconfig_webhook.go:
```diff
@@ -0,0 +1,78 @@
+	}
+}
+
+//+kubebuilder:webhook:path=/validate-redhatcop-redhat-io-v1alpha1-ldapauthengineconfig,mutating=false,failurePolicy=fail,sideEffects=None,groups=redhatcop.redhat.io,resources=ldapauthengineconfigs,verbs=create;update,versions=v1alpha1,name=vldapauthengineconfig.kb.io,admissionReviewVersions=v1
```
please remove the create event here.

erlisb replied:
removed.

raffaelespazzoli commented on docs/auth-engines.md:
```diff
@@ -92,3 +93,104 @@ This CR is roughly equivalent to this Vault CLI command:
+    serviceAccount:
+      name: admin-sa
+  bindDN: cn=vault,ou=Users,dc=example,dc=com
+  bindPass: xxxxxxxxxxxxx
```
this should be passed as a secret, please refer to other engines that do similar things

erlisb replied:
I try to do it same way as DatabaseSecretEngineConfig type. CRDs are generated correctly, but the bindDN & bindPass fields somehow do not get populated : 

```
  bindCredentials:
    passwordKey: password
    secret:
      name: bindcredentials
    usernameKey: username
  bindDN: ""
  bindPass: ""
```

raffaelespazzoli replied:
you have to initialize those fields in the `PrepareInternalValues` fuinction. `bindPass`, should not be exposed by the CRD.
Let me know if you need more help.
see also: https://github.com/redhat-cop/vault-config-operator/blob/7edf4ffb162efa3d4232c98c604146f40bf839d2/api/v1alpha1/databasesecretengineconfig_types.go#L73

erlisb replied:
Hi @raffaelespazzoli thank you for the feedback.
tbh I followed the same approach as _DatabaseSecretEngineConfig_ type and set both methods _PrepareInternalValues_ & _setInternalCredentials_ accordingly. Did a few more changes, but still I am not able to retrieve those values from _bindCredentials_ property.

raffaelespazzoli replied:
I am not sure what you mean by "retrieving those credentials". You really are not supposed to retrieve them. Can you explain what you are trying to do? Vault will not show them back to you.

erlisb replied:
Hi @raffaelespazzoli, I thought that when creating the CR for _LDAPAuthEngineConfig_, the values for _bindDN_ & _bindPass_ would get populated by _bindCredentials_, but instead I get : 

provided CR : 

```
  url: "ldaps://ldap-xxx-a.xxx.xxx.xxx,ldaps://ldap-xxx-b.xxx.xxx.xxx,ldaps://ldap-xxx-c.xxx.xxx.xxx"
  caseSensitiveNames: false
  bindDN: "supersecret"
  bindCredentials:
    secret: 
      name: bindcredentials
    usernameKey: username
    passwordKey: password
  userDN: "ou=Users,ou=XXXX,ou=Company,dc=xxx,dc=xxx,dc=xxx" 
```

populated CR (after creation): 

```
    authentication:
      path: kubernetes
      role: policy-admin
      serviceAccount:
        name: default
    bindCredentials:
      passwordKey: password
      secret:
        name: bindcredentials
      usernameKey: username
    bindDN: supersecret
    bindPass: ""
    caseSensitiveNames: false
```
Since these two properties (_bindDN_ & _bindPass_) are also part of the payload send to Vault API, I though these should be somehow visible.

raffaelespazzoli replied:
mm, I'm lost...I don't understand what you are trying to do, the "supersecret" value in bindDN is confusing. Anyway, `bindPass` is clearly a secret and should NOT appear in the CR. Looking at your code, it seems to be the case. I don't know LDAP in detail, if `bindDN` is also a secret, you should treat it the same way as `bindPass`. If not, what you have in the first code quote seems correct.
I don't understand what the second quote is supposed to be.., is that what you would expect to see?

erlisb replied:
Hi @raffaelespazzoli , sorry for the confusion. The bindDN field could or not be a secret. This is optional and up to the user. Regarding the bindPass, you are correct, it shouldn't appear in the CR. 
Exactly, the second quote is what I thought would be the correct one, but it was not the case, as you pointed out.
Do you have any suggestion how I can test this ? Should I spin up a Minishift/CodeReady/kind cluster + OpenLDAP ? I can't test it at the moment on a OpenShift Prod Cluster.
thnx

raffaelespazzoli replied:
Yes, we try to test our stuff with a real integration test before releasing it. At a minimum script the test so it can be execute manually. Better would be to automate it, given that now we can do it with this operator.
A while ago a scripted how to deploy an LDAP server on OCP and populate with some test data. Maybe you can start from here: https://cloud.redhat.com/blog/orgs-management-and-team-onboarding-in-openshift-a-fully-automated-approach.

erlisb replied:
Hi @raffaelespazzoli, 
I took some time to study your repo for `orgs-management-ocp/team-onboarding` topic you provided. I am still working to automate the integration test. Meanwhile, I am doing some manual tests : 

1. Spin-up a `kind `cluster where both `Vault ` (`dev` mode) and `LDAP ` are deployed (they can see each other in terms of network)
2. Create an `AuthEngineMount `CR named  `ldap-test`
3. Create an `LDAPAuthEngineConfig` CR as follows : 

```
  path: ldap-test
  url: "ldap://ldap.ldap.svc.cluster.local"
  caseSensitiveNames: false
  bindDN: "cn=admin,cn=config"
  userDN: "dc=example,dc=com"
  bindCredentials:
    secret: 
      name: bindcredentials
    usernameKey: username
    passwordKey: password
  insecureTls: true  ---> TLS has been disabled also in LDAP side
```
It seems that at least the `bindCredentials` is working okay, since I can login there, but I get the following errors : 

`2022-04-29T15:01:30.536Z [DEBUG] auth.ldap.auth_ldap_9953873a: error getting user bind DN: error="LDAP search for binddn failed: LDAP Result Code 32 \"No Such Object\": "`

if that was not the case, we would get : 

`022-04-29T14:03:09.371Z [DEBUG] auth.ldap.auth_ldap_9953873a: error getting user bind DN: error="LDAP search for binddn failed: LDAP bind (service) failed: LDAP Result Code 49 \"Invalid Credentials\": "`

Since I don't know LDAP in detail, I am not sure what I am missing  in the LDAP config (CR).
I am using the LDAP test data part of your repo. Login in Vault (via ldap) :  `username`: trevor , `password `:  xxxxx

erlisb replied:
UPDATE: 
Apparently, I was using the wrong `bindDN `& `baseDN`.
I used the following config (CR) instead : 

```
  path: ldap-test
  url: "ldap://ldap.ldap.svc.cluster.local"
  caseSensitiveNames: false
  bindDN: "cn=admin,dc=example,dc=com"
  userDN: "ou=users,dc=example,dc=com"
  bindCredentials:
    secret: 
      name: bindcredentials
    usernameKey: username
    passwordKey: password
  insecureTls: true
```

and I am able to login with username/uid : `trevor `and its password.

Vault & LDAP logs : 

Vault 

```
2022-05-02T21:06:25.566Z [DEBUG] auth.ldap.auth_ldap_ee620879: compiling search filter: search_filter="({{.UserAttr}}={{.Username}})"
2022-05-02T21:06:25.566Z [DEBUG] auth.ldap.auth_ldap_ee620879: discovering user: userdn="ou=users,dc=example,dc=com" filter="(cn=trevor)"
2022-05-02T21:06:25.567Z [DEBUG] auth.ldap.auth_ldap_ee620879: user binddn fetched: username=trevor binddn="uid=trevor,ou=Users,dc=example,dc=com"
2022-05-02T21:06:25.567Z [DEBUG] auth.ldap.auth_ldap_ee620879: re-bound to original binddn
2022-05-02T21:06:25.567Z [WARN]  auth.ldap.auth_ldap_ee620879: groupfilter is empty, will not query server
2022-05-02T21:06:25.567Z [DEBUG] auth.ldap.auth_ldap_ee620879: groups fetched from server: num_server_groups=0 server_groups=[]
2022-05-02T21:06:25.567Z [DEBUG] auth.ldap.auth_ldap_ee620879: compiling search filter: search_filter="({{.UserAttr}}={{.Username}})"
2022-05-02T21:06:25.568Z [DEBUG] auth.ldap.auth_ldap_ee620879: discovering user: userdn="ou=users,dc=example,dc=com" filter="(cn=trevor)"
```

LDAP

```
62704834 conn=1030 fd=12 ACCEPT from IP=10.42.0.178:56588 (IP=0.0.0.0:389)
62704834 conn=1030 op=0 BIND dn="cn=admin,dc=example,dc=com" method=128
62704834 conn=1030 op=0 BIND dn="cn=admin,dc=example,dc=com" mech=SIMPLE ssf=0
62704834 conn=1030 op=0 RESULT tag=97 err=0 text=
62704834 conn=1030 op=1 SRCH base="ou=users,dc=example,dc=com" scope=2 deref=0 filter="(cn=trevor)"
62704834 conn=1030 op=1 SRCH attr=cn
62704834 <= mdb_equality_candidates: (cn) not indexed
62704834 conn=1030 op=1 SEARCH RESULT tag=101 err=0 nentries=1 text=
62704834 conn=1030 op=2 BIND anonymous mech=implicit ssf=0
62704834 conn=1030 op=2 BIND dn="uid=trevor,ou=Users,dc=example,dc=com" method=128
62704834 conn=1030 op=2 BIND dn="uid=trevor,ou=Users,dc=example,dc=com" mech=SIMPLE ssf=0
62704834 conn=1030 op=2 RESULT tag=97 err=0 text=
62704834 conn=1030 op=3 BIND anonymous mech=implicit ssf=0
62704834 conn=1030 op=3 BIND dn="cn=admin,dc=example,dc=com" method=128
62704834 conn=1030 op=3 BIND dn="cn=admin,dc=example,dc=com" mech=SIMPLE ssf=0
62704834 conn=1030 op=3 RESULT tag=97 err=0 text=
62704834 conn=1030 op=4 BIND anonymous mech=implicit ssf=0
62704834 conn=1030 op=4 BIND dn="cn=admin,dc=example,dc=com" method=128
62704834 conn=1030 op=4 BIND dn="cn=admin,dc=example,dc=com" mech=SIMPLE ssf=0
62704834 conn=1030 op=4 RESULT tag=97 err=0 text=
62704834 conn=1030 op=5 SRCH base="ou=users,dc=example,dc=com" scope=2 deref=0 filter="(cn=trevor)"
62704834 conn=1030 op=5 SRCH attr=cn
62704834 <= mdb_equality_candidates: (cn) not indexed
62704834 conn=1030 op=5 SEARCH RESULT tag=101 err=0 nentries=1 text=
62704834 conn=1030 fd=12 closed (connection lost)
```

Do you think that this is sufficient for the integration test ?
Thnx

raffaelespazzoli replied:
yes I think this is good enough

erlisb replied:
Do I need to do anymore here, before we set this to "Resolved" ?

raffaelespazzoli commented on docs/auth-engines.md:
```diff
@@ -92,3 +93,104 @@ This CR is roughly equivalent to this Vault CLI command:
+  groupFilter:
+  insecureTls: true
+  path: ldap
+  tokenReviewerServiceAccount:
```
I don't understand what this is used for, can you explain? it looks like it's not a field of the LDAP secret engine API.

erlisb replied:
I removed this. I thought it was used in the Authentication part, but I was mistaken. Documentation has been updated accordingly.

raffaelespazzoli commented on test/kv-secret-engine.yaml:
```diff
@@ -5,6 +5,6 @@ metadata:
   authentication: 
     path: kubernetes
-    role: kv-engine-admin
+    role: policy-admin
```
can you revert this change, it should not be needed

erlisb replied:
change has been reverted.

raffaelespazzoli commented:
@erlisb is this ready for a final check, or are there more test you need to run?

erlisb commented:
@raffaelespazzoli, I did several tests locally. Seems to be okay.
I added also a `ldap-setup` step in the makefile. What else is needed ?

raffaelespazzoli reviewed (changes requested):
I added a bunch on nitpicks, sorry for not spotting them before. The I have two questions:
1. I don't see a mapping 1 to 1 between the fiuelds defined in the config CRD and the field described in the API doc (https://www.vaultproject.io/api-docs/secret/ad#configuration). Can you share the info you used?
2. why didn't you also map the role api? I get not mapping the check-in/out and library API, but what about the role. Would it make sense to do it?

erlisb commented:
> I added a bunch on nitpicks, sorry for not spotting them before. The I have two questions:
> 
> 1. I don't see a mapping 1 to 1 between the fiuelds defined in the config CRD and the field described in the API doc (https://www.vaultproject.io/api-docs/secret/ad#configuration). Can you share the info you used?
> 2. why didn't you also map the role api? I get not mapping the check-in/out and library API, but what about the role. Would it make sense to do it?

> I added a bunch on nitpicks, sorry for not spotting them before. The I have two questions:
> 
> 1. I don't see a mapping 1 to 1 between the fiuelds defined in the config CRD and the field described in the API doc (https://www.vaultproject.io/api-docs/secret/ad#configuration). Can you share the info you used?
> 2. why didn't you also map the role api? I get not mapping the check-in/out and library API, but what about the role. Would it make sense to do it?



> I added a bunch on nitpicks, sorry for not spotting them before. The I have two questions:
> 
> 1. I don't see a mapping 1 to 1 between the fiuelds defined in the config CRD and the field described in the API doc (https://www.vaultproject.io/api-docs/secret/ad#configuration). Can you share the info you used?
> 2. why didn't you also map the role api? I get not mapping the check-in/out and library API, but what about the role. Would it make sense to do it?

@raffaelespazzoli thanks for your review.
regarding 1. I used this [info ](https://www.vaultproject.io/api-docs/auth/ldap) and set every field to camel case (followed same conventions as in the KubernetesAuthEngineConfig)
regarding 2. I don't get what you mean by `role` ? I thought a `role` was not needed in this Auth method. Are you referring to LDAP [Groups](https://www.vaultproject.io/api-docs/auth/ldap#create-update-ldap-group) & [Users](https://www.vaultproject.io/api-docs/auth/ldap#create-update-ldap-user) ? If yes, I was considering this for an upcoming PR.

erlisb commented on api/v1alpha1/ldapauthengineconfig_types.go:
```diff
@@ -0,0 +1,402 @@
+	// StartTls If true, issues a StartTLS command after establishing an unencrypted connection.
+	// +kubebuilder:validation:Optional
+	// +kubebuilder:default=false
+	StartTls bool `json:"startTls"`
```
Done

erlisb commented on api/v1alpha1/ldapauthengineconfig_types.go:
```diff
@@ -0,0 +1,402 @@
+	// TlsMinVersion Minimum TLS version to use. Accepted values are tls10, tls11, tls12 or tls13
+	// +kubebuilder:validation:Optional
+	// +kubebuilder:default="tls12"
+	TlsMinVersion string `json:"tlsMinVersion"`
```
Done

erlisb commented on api/v1alpha1/ldapauthengineconfig_types.go:
```diff
@@ -0,0 +1,402 @@
+	// InsecureTls If true, skips LDAP server SSL certificate verification - insecure, use with caution!
+	// +kubebuilder:validation:Optional
+	// +kubebuilder:default=false
+	InsecureTls bool `json:"insecureTls"`
```
Done

erlisb commented on api/v1alpha1/ldapauthengineconfig_types.go:
```diff
@@ -0,0 +1,402 @@
+	// ClientTlsCert Client certificate to provide to the LDAP server, must be x509 PEM encoded
+	// +kubebuilder:validation:Optional
+	// +kubebuilder:default=""
+	ClientTlsCert string `json:"clientTlsCert,omitempty"`
```
Done

erlisb commented on api/v1alpha1/ldapauthengineconfig_types.go:
```diff
@@ -0,0 +1,402 @@
+	// ClientTlsKey Client certificate key to provide to the LDAP server, must be x509 PEM encoded
+	// +kubebuilder:validation:Optional
+	// +kubebuilder:default=""
+	ClientTlsKey string `json:"clientTlsKey,omitempty"`
```
Done

erlisb commented on api/v1alpha1/ldapauthengineconfig_types.go:
```diff
@@ -0,0 +1,402 @@
+	// TokenTtl The incremental lifetime for generated tokens. This current value of this will be referenced at renewal time.
+	// +kubebuilder:validation:Optional
+	// +kubebuilder:default=""
+	TokenTtl string `json:"tokenTtl,omitempty"`
```
Done

erlisb commented on api/v1alpha1/ldapauthengineconfig_types.go:
```diff
@@ -0,0 +1,402 @@
+	// TokenMaxTtl The maximum lifetime for generated tokens. This current value of this will be referenced at renewal time
+	// +kubebuilder:validation:Optional
+	// +kubebuilder:default=""
+	TokenMaxTtl string `json:"tokenMaxTtl,omitempty"`
```
Done

erlisb commented on api/v1alpha1/ldapauthengineconfig_types.go:
```diff
@@ -0,0 +1,402 @@
+	// TokenExplicitMaxTtl If set, will encode an explicit max TTL onto the token. This is a hard cap even if token_ttl and token_max_ttl would otherwise allow a renewal.
+	// +kubebuilder:validation:Optional
+	// +kubebuilder:default=""
+	TokenExplicitMaxTtl string `json:"tokenExplicitMaxTtl,omitempty"`
```
Done

erlisb commented on api/v1alpha1/webhook_suite_test.go:
```diff
@@ -124,6 +124,9 @@ var _ = BeforeSuite(func() {
 	err = admissionv1beta1.AddToScheme(scheme)
 	Expect(err).NotTo(HaveOccurred())
 
+	err = admissionv1beta1.AddToScheme(scheme)
```
tbh, I don't know what that is used for ?

raffaelespazzoli replied:
please remove it since it does the same thing a the previous lines.

erlisb replied:
fixed. There are several copies of the same lines at 82-125

raffaelespazzoli commented:
ok, so I was confusing the auth engine with the secret engine, my bad. What you have is good sorry.
There is one last comment open. Also commit (https://github.com/redhat-cop/vault-config-operator/pull/51/commits/f9595dd369f50bd079c3ee322befaea3b100671b) didn't seem necessary, maybe you want to undo it. 
Then we are ready to go I think.
